src/Lykke.Job.CandlesHistoryWriter.Services/HistoryMigration/TradesMigrationService.cs

```diff
@@ -0,0 +1,185 @@
+                            // To avoid looking through the whole trades batch several times, we derive the
+                            // bigger time intervals from the smaller ones. For example, let's assume that we
+                            // generate 1000 sec candles from our batch of 2381 trades. Then, for calculating
+                            // minute candles we need to iterate only 1000 second candles, and this will give
+                            // us 16 minute candles. Next, we iterate 16 minute candles to get 1 hour candle,
+                            // and so on, instead of looking through 2381 trades for each time interval.
+                            foreach (var interval in Candles.Constants.StoredIntervals)
+                            {
+                                // It's important for Constants.StoredIntervals to be ordered by time period increase,
+                                // because we will calculate candles for each period based on previous period candles.
+                                var currentCandles = interval == CandleTimeInterval.Sec
+                                    ? new TradesCandleBatch(searchToken.AssetPairId, searchToken.SearchToken, searchToken.ReverseSearchToken, interval, tradesBatch)
+                                    : new TradesCandleBatch(searchToken.AssetPairId, interval, smallerCandles);
+
+                                ExtendStoredCandles(currentCandles);
+
+                                _tradesMigrationHealthService[searchToken.AssetPairId].SummarySavedCandles +=
+                                    currentCandles.Candles.Count;
+
+                                // We can not derive month candles from weeks for they may lay out of the borders
+                                // of the month. While generating a month candles, we should use day candles as a
+                                // data source instead.
+                                if (interval != CandleTimeInterval.Week)
+                                    smallerCandles = currentCandles;
+
+                                // Preffered to perform inserting in parralel style.
+                                batchInsertTasks.Add(
+                                    _candlesHistoryRepository.InsertOrMergeAsync(
+                                        currentCandles.Candles.Values,
+                                        currentCandles.AssetId,
+                                        TradesCandleBatch.PriceType,
+                                        currentCandles.TimeInterval));
+                            }
+
+                            _tradesMigrationHealthService[searchToken.AssetPairId].SummaryFetchedTrades += batchCount;
+
+                            await Task.WhenAll(batchInsertTasks.ToArray());
+
+                            await _log.WriteInfoAsync(nameof(TradesMigrationManager), nameof(MigrateTradesCandlesAsync),
```

Comment:
You could use https://github.com/LykkeCity/CommonDotNetLibraries/blob/master/src/Lykke.Common/Log/LogExtensions.cs#L19 in every class to avoid repetiotion of the component in every _log.WriteXXX call